contracts/modules/cover/Cover.sol

```diff
@@ -0,0 +1,570 @@
+import "../../abstract/MasterAwareV2.sol";
+import "../../interfaces/IMemberRoles.sol";
+import "hardhat/console.sol";
+
+
+contract Cover is ICover, ERC721, MasterAwareV2 {
+
+  Cover[] public override covers;
+  mapping(uint => StakingPool[]) stakingPoolsForCover;
+
+  Product[] public override products;
+  mapping(uint => uint) capacityFactors;
+
+  ProductType[] public override productTypes;
+
+  mapping(uint => uint) initialPrices;
+
+  mapping(uint => uint96) public override activeCoverAmountInNXM;
+
+  /*
+   (productId, poolAddress) => lastPrice
+   Last base prices at which a cover was sold by a pool for a particular product.
+  */
+  mapping(uint => mapping(address => uint)) lastPrices;
+
+  /*
+   (productId, poolAddress) => lastPriceUpdate
+   Last base price update time.
+  */
+  mapping(uint => mapping(address => uint)) lastPriceUpdate;
+
+
+  /* === CONSTANTS ==== */
+
+  uint public REWARD_BPS = 5000;
+  uint public constant PERCENTAGE_CHANGE_PER_DAY_BPS = 100;
+  uint public constant BASIS_PRECISION = 10000;
+  uint public constant STAKE_SPEED_UNIT = 100000e18;
+  uint public constant PRICE_CURVE_EXPONENT = 7;
+  uint public constant MAX_PRICE_PERCENTAGE = 1e20;
+
+  /* === MODIFIERS ==== */
+
+  modifier onlyAdvisoryBoard {
+    uint abRole = uint(IMemberRoles.Role.AdvisoryBoard);
+    require(
+      memberRoles().checkRole(msg.sender, abRole),
+      "Cover: Caller is not an advisory board member"
+    );
+    _;
+  }
+
+  constructor(string memory name_, string memory symbol_) ERC721(name_, symbol_) {
+  }
+
+  /* === MUTATIVE FUNCTIONS ==== */
+
+  function buyCover(
+    address owner,
+    uint24 productId,
+    uint8 payoutAsset,
+    uint96 amount,
+    uint32 period,
+    uint maxPrice,
+    StakingPool[] memory stakingPools
+  ) external payable override returns (uint /*coverId*/) {
+    require(initialPrices[productId] != 0, "Cover: product not initialized");
+
+    (uint coverId, uint priceInAsset) = _createCover(owner, productId, payoutAsset, amount, period, stakingPools);
+    require(priceInAsset <= maxPrice, "Cover: Price exceeds maxPrice");
+    retrievePayment(priceInAsset, payoutAsset);
+    return coverId;
+  }
+
+  function _createCover(
+    address owner,
+    uint24 productId,
+    uint8 payoutAsset,
+    uint96 amount,
+    uint32 period,
+    StakingPool[] memory stakingPools
+  ) internal returns (uint coverId, uint premiumInAsset) {
+
+    // convert to NXM amount
+    uint amountLeftToCoverInNXM = uint(amount) * 1e18 / pool().getTokenPrice(pool().assets(payoutAsset));
+    activeCoverAmountInNXM[productId] += uint96(amountLeftToCoverInNXM);
+
+    uint totalPremiumInNXM = 0;
+
+    for (uint i = 0; i < stakingPools.length; i++) {
+      if (amountLeftToCoverInNXM == 0) {
+        break;
+      }
+
+      IStakingPool stakingPool = IStakingPool(stakingPools[i].poolAddress);
+      (uint coveredAmount, uint premiumInNXM) = buyCoverFromPool(stakingPool, productId, amountLeftToCoverInNXM, period);
```

Comment:
This currently takes the max amount of each pool instead of using the input and only buying a specified amount from each pool. However, when purchasing a limited capacity out of each pool, if a specified pool doesn't have enough it might revert the whole thing. We'll need to find a middle-ground here.

Reply:
we discussed having some sort of slippage for amount, but it seemed to me it's getting complicated when combined with the price slippage, plus it would be an edge case that occurs only when you're at the upper limit of capacity

Reply:
discussed offline